### PR TITLE
Explore: Reorganize useStateSync code

### PR DESCRIPTION
**Pull request by ifrost**
No actual changes in the behavior, just moving things around to improve the readability of a quite big (465-lines long) `useStateSync/index.ts`. The code in the file was already organized nicely, but it may be hard to follow all the flows as everything is wrapped in arrow functions and useEffect.

With this change, the main entry point `useStateSync` is reduced to mainly:

```typescript
export function useStateSync(params: ExploreQueryParams) {

  // ...

  useEffect(() => {
    const unsubscribe = dispatch(
      addListener({
        predicate: (action) => syncToURLPredicate(paused, action),
        effect: async (_, { cancelActiveListeners, delay, getState }) => {
          // ...
          syncToURL(getState, prevParams, initState, location);
        },
      })
    );

    return () => unsubscribe();
  }, [dispatch, location]);

  useEffect(() => {
    // ...
    if (!initialized) {
      initializeFromURL(urlState, initState, orgId, dispatch, location);
    }

    if (fullyInitialized) {
      syncFromURL(urlState, panesState, dispatch);
    }
  }, [dispatch, panesState, orgId, location, params, warning]);
}
```

Which should be easier to digest.

The code was split into following files:

```
/useStateSync
  /synchronizer
    /fromURL.ts      --> functions to convert URL to Explore State
    /toURL.ts        --> functions to convert Explore State to URL
    /init.ts         --> functions to initialize Explore State from URL
  .external.utils.ts --> URL-management related tools we expose
  .internal.utils.ts --> URL-management related tools used by synchronizers above
```

### CODE REVIEW

File: public/app/features/explore/hooks/useStateSync/synchronizer/init.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  getPaneDatasource,
+  getQueryFilter,
+  isMixedDatasource,
+  removeQueriesWithInvalidDatasource,
+} from '../internal.utils';
+import { ExploreURLV1 } from '../migrators/v1';
+
+export function initializeFromURL(
+  urlState: ExploreURLV1,
+  initState: MutableRefObject<'pending' | 'done' | 'notstarted'>,
```
Comment (Elfo404) [inline public/app/features/explore/hooks/useStateSync/synchronizer/init.ts:26]: given we are now breaking this into multiple files, what about exporting a type somewhere like

```ts
type InitState = 'notstarted' | 'pending' | 'done';
```

to use here and in `public/app/features/explore/hooks/useStateSync/synchronizer/toURL.ts` as `MutableRefObject<InitState>` ?

**Review by Elfo404 — APPROVED** [review]
Works as expected, great stuff!

just 2 nits, but LGTM

File: public/app/features/explore/hooks/useStateSync/index.ts [hunk 2]
```diff
@@ -47,65 +38,13 @@ export function useStateSync(params: ExploreQueryParams) {
...
-              (!!prevParams.current.panes && Object.values(prevParams.current.panes).filter(Boolean).length === 0) ||
-              initState.current === 'pending';
-
-            prevParams.current = {
-              panes: JSON.stringify(panesQueryParams),
-            };
-
-            location.partial({ panes: prevParams.current.panes }, replace);
-          }
+          syncToURL(getState, prevParams, initState, location);
```
[the remark this answers is not in the document]
Comment (ifrost) [reply public/app/features/explore/hooks/useStateSync/index.ts]: Fixed in 10a31cdeaf3015dd9eacaf21beb5b22c08bb26ec

File: public/app/features/explore/hooks/useStateSync/synchronizer/init.ts (hunk 1 above)
Comment (ifrost) [reply public/app/features/explore/hooks/useStateSync/synchronizer/init.ts]: Good point! Updated in 1586cc5e02e2841cd26cbdd0e9134ba17360959f.
